healthcheck/health.go

```diff
@@ -66,3 +73,96 @@ func HealthCheck(session *mgo.Session, okMemberStates []status.MemberState) (Sta
 
 	return StateFailed, state, fmt.Errorf("member has unhealthy replication state: %s", state)
 }
+
+func HealthCheckLiveness(session *mgo.Session, okMemberStates []status.MemberState, startupDelaySeconds int64) (State, *status.MemberState, error) {
+	isMasterResp := IsMasterResp{}
+	err := session.Run(bson.D{{Name: "isMaster", Value: 1}}, &isMasterResp)
+	if err != nil {
+		return StateFailed, nil, fmt.Errorf("isMaster returned error %v", err)
+	}
+	if isMasterResp.Ok == 0 {
+		return StateFailed, nil, errors.New(isMasterResp.Errmsg)
+	}
+
+	info, err := session.BuildInfo()
+	if err != nil {
+		return StateFailed, nil, fmt.Errorf("failed to get mongo build info: %v", err)
+	}
+
+	replSetStatusCommand := bson.D{{Name: "replSetGetStatus", Value: 1}}
+	if info.Version < "4.2.1" {
+		// https://docs.mongodb.com/manual/reference/command/replSetGetStatus/#syntax
+		replSetStatusCommand = append(replSetStatusCommand, bson.DocElem{Name: "initialSync", Value: 1})
+	}
+
+	replSetGetStatusResp := ReplSetStatus{}
+	err = session.Run(replSetStatusCommand, &replSetGetStatusResp)
+	if err != nil {
+		return StateFailed, nil, fmt.Errorf("replSetGetStatus returned error %v", err)
+	}
+
+	err = replSetGetStatusResp.CheckState(startupDelaySeconds)
+	if err != nil {
+		return StateFailed, &replSetGetStatusResp.MyState, err
+	}
+
+	return StateOk, &replSetGetStatusResp.MyState, nil
+}
+
+func checkServerStatus(session *mgo.Session) error {
+	status := &ServerStatus{}
+	err := session.DB("admin").Run(bson.D{{Name: "serverStatus", Value: 1}}, status)
+	if err != nil {
+		return err
+	}
+	if status.Ok == 0 {
+		return errors.New(status.Errmsg)
+	}
+	return nil
+}
+
+type ServerStatus struct {
+	Ok     int    `bson:"ok" json:"ok"`
+	Errmsg string `bson:"errmsg,omitempty" json:"errmsg,omitempty"`
+}
+
+type IsMasterResp struct {
+	IsMaster bool `bson:"ismaster" json:"ismaster"`
+
+	Ok     int    `bson:"ok" json:"ok"`
+	Errmsg string `bson:"errmsg,omitempty" json:"errmsg,omitempty"`
+}
+
+type ReplSetStatus struct {
+	status.Status     `bson:",inline"`
+	InitialSyncStatus InitialSyncStatus `bson:"initialSyncStatus" json:"initialSyncStatus"`
+}
+
+type InitialSyncStatus interface{}
+
+func (rs ReplSetStatus) CheckState(startupDelaySeconds int64) error {
+	if rs.Ok == 0 {
+		return errors.New(rs.Errmsg)
+	}
+
+	uptime := rs.GetSelf().Uptime
+
+	switch rs.MyState {
+	case status.MemberStatePrimary, status.MemberStateSecondary, status.MemberStateArbiter:
+		return nil
+	case status.MemberStateStartup, status.MemberStateStartup2:
+		if (rs.InitialSyncStatus == nil && uptime > 30) || (rs.InitialSyncStatus != nil && uptime > startupDelaySeconds) {
```

Comment:
opLogSize is a factor in calculating slow startup time. The full oplog is scanned on startup. When you have a large oplog (say 100GB+ on spinning magnetic disk) the startup can take > 1 min.

You removed this optimization (for calculating max bound of startup time) from earlier versions of algorithm, for simplicity. As it would be 4 x 30s = 2 mins before the kill would happen anyhow.

I.i.r.c I wanted to allow max(1 min, 1 min * (oplogSize / 100GB) )

Reply:
@akira-kurogane can we simplify this case to this formula? (don't check rs.InitialSyncStatus at all)
```suggestion
		if uptime > startupDelaySeconds {
```

Reply:
We should allow a long time (even an hour or two) for initial syncs to complete. The presence of this sub-object "initialSyncStatus" in the replSetGetStatus is how we detect that.

Reply:
Hi Mykola - as we discussed in the chat just now we should not restart while the initialSyncStatus field is present.